remill/BC/Annotate.h

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright (c) 2019 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Metadata.h>
+
+#include "remill/BC/Version.h"
+
+#include <glog/logging.h>
+
+namespace remill {
+
+// For all function holds that OriginType should have two static strings
+// * metadata_kind that is the identifier (kind) of metadata
+// * metadata_value that is the value that identifies the OriginType itself
+
+
+struct BaseFunction {
+  static const std::string metadata_kind;
+  // Each class that inherits from some parent class appends its own metadata_value string with dot
+  // to the parents one, therefore it is possible for search to easily include all subclasses as well
+  static const std::string metadata_value;
+};
+
+// Unfortunately we must define the values of static vars in .cpp file
+#define DECLARE_FUNC_ORIGIN_TYPE(children, parent) \
+struct children : parent { \
+  static const std::string metadata_value; \
+}
+
+DECLARE_FUNC_ORIGIN_TYPE( LiftedFunction, BaseFunction );
+DECLARE_FUNC_ORIGIN_TYPE( EntrypointFunction, BaseFunction );
+DECLARE_FUNC_ORIGIN_TYPE( ExternalFunction, BaseFunction );
+DECLARE_FUNC_ORIGIN_TYPE( AbiLibraries, ExternalFunction );
+DECLARE_FUNC_ORIGIN_TYPE( CFGExternal, ExternalFunction );
+DECLARE_FUNC_ORIGIN_TYPE( Helper, BaseFunction );
+DECLARE_FUNC_ORIGIN_TYPE( RemillHelper, Helper );
+DECLARE_FUNC_ORIGIN_TYPE( McSemaHelper, Helper );
+
+#undef DECLARE_FUNC_ORIGIN_TYPE
+
+// Default kind to be used, user is free to choose different, for example if each function
+// is supposed to be part of multiple pairs
+const std::string TieKind = "remill.function.tie";
+
+// Note(Aiethel): I tried to make them static methods and while it works it is not really worth
```

Comment:
What do you mean "tried"? How did you measure worthiness? The value of `static`, or `inline static` for definitions in headers is so that duplicate definition errors don't creep in when these things are used.